controllers/front/OrderConfirmationController.php

```diff
@@ -65,9 +67,18 @@ public function init()
         if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
             Tools::redirect($redirectLink);
         }
-        $module = Module::getInstanceById((int) ($this->id_module));
-        if ($order->module != $module->name) {
-            Tools::redirect($redirectLink);
+
+        // free order uses -1 as id_module, it has a special check here
```

Comment:
Do you have an idea where the -1 is settled?

Reply:
checkFreeOrder does a redirect back here. 

The original request has just free_order=1 as a parameter. 
The first check above is to call checkFreeOrder() (which should probably be renamed to handleFreeOrder() ) 
In checkFreeOrder method line 177 (https://github.com/PrestaShop/PrestaShop/pull/22389/files#diff-b8ddf176f2937afab07da250219b256a0a9c06b1fd8f2901797f1f0a2d85b34cR177) does basket => order and redirects back here (just like a real payment method module would) and the process continues as usual.

I don't really like that -1 but it helps with the line 67 check that wants "true" value for module and we cannot use any positive integer as that would try to load some random module..

The real fix should probably be to make a "freePayment" module that cannot be uninstalled and that does a "real" payment process just like all the other payment modules...

Reply:
I don't think the `id_module` needs to be used here. If you need something special you can add a custom dedicated parameter that will be more evident to use and understand.
It's the same with `$order->module` as `free_order`.
If it's not a module we should bypass checks and should do only what we need.

Reply:
I have not dug too deep into all this logic so this is also a change that should probably be checked by few people who are well versed on the payment process and all the edge cases (especially from security point of view). 

As far as I know this is safe (the module name check should make sure that basket with "paypal" should not go through with free_order=1 etc), that said I'm not 100% sure (I've been reading prestashop code for just a few days now...)

Reply:
You're right about security, it should be up to the payment module to provide free order.

Reply:
@PierreRambaud yup, another flag would work also and will probably be clearer.

Do you know why the orderpresenter is initiated in this init function? If that could be moved to initContent the flag could just return from init which would make this pretty clear and the check on line 63 would not have be touched either.

Reply:
If I had to guess none of the payment modules handle 0 total orders correctly as it is handled internally with https://github.com/PrestaShop/PrestaShop/blob/develop/classes/PaymentFree.php . At the moment they cannot even test the situation where the order is 0 total (no payments can be selected and the form goes to ?free_order=1 and gets handled very differently).

But if that PaymentFree class could be "raised" to be a real module and payment selection forces that module to be only one available when order has 0 total the process would be the same for all orders and every payment module developer can keep ignoring the 0 price situation.

That raises new question if that new PaymentFree should always be available (aka, no uninstall available for it). Or if it is installed should the payment selector just show a "No payment methods available for 0 priced orders". Or maybe something else completely?

Reply:
I will try to investigate on Monday :+1:

Reply:
Still didn't have the time to investigate, sorry :(

Reply:
@PierreRambaud no worries, aren't we all still just eating christmas leftovers and waiting for 2020 to end :D

Reply:
@tswfi True story :roll_eyes:

Reply:
Hey, finally took time to investigate, Tried to change many things to make sure I can't bypass and everything seems ok :+1:

Reply:
Thanks :)